settings.org
#+title: Emacs Config
#+author: Amo DelBello
#+startup: content

* Config
** Variables and Modes
#+begin_src emacs-lisp
  (use-package emacs :ensure nil
    :config
    ;; Use C-h for delete-backward-char
    (global-set-key (kbd "C-h") 'delete-backward-char)

    ;; Set custom prefix key ("C-z")
    (define-prefix-command 'z-map)
    (global-set-key (kbd "C-z") 'z-map)

    (global-display-line-numbers-mode 0)
    (fset 'yes-or-no-p 'y-or-n-p)
    (recentf-mode 1)
    (savehist-mode 1)
    (whitespace-mode -1)
    (delete-selection-mode 1)
    (blink-cursor-mode 0)
    (electric-pair-mode 1)
    (turn-on-visual-line-mode)
    (visual-line-mode 1)
    (which-key-mode 1)
    (put 'narrow-to-region 'disabled nil)

    (setq save-interprogram-paste-before-kill t
          backup-directory-alist '(("." . (dot-env-get 'BACKUP_DIRECTORY "~/.emacs.d/backups")))
          text-mode-ispell-word-completion nil
          read-extended-command-predicate #'command-completion-default-include-p
          initial-major-mode 'org-mode
          initial-scratch-message "This buffer is for text that is not saved.\nTo create a file, visit it with \\[find-file] and enter text in its buffer.\n\n"
          calendar-week-start-day 1
          confirm-kill-emacs #'y-or-n-p
          desktop-save-mode nil
          vc-follow-symlinks nil
          require-final-newline t
          enable-recursive-minibuffers t
          python-indent-guess-indent-offset-verbose nil
          custom-safe-themes t
          calc-window-height 10
          ispell-program-name (dot-env-get 'ISPELL_PATH "/usr/bin/ispell")
          dictionary-server "dict.org"
          warning-minimum-level :error
          default-directory (dot-env-get 'DEFAULT_DIRECTORY "~/")
          scroll-step 1
          compilation-window-height 24
          frame-inhibit-implied-resize t
          recentf-max-menu-items 25
          recentf-max-saved-items 25
          inferior-lisp-program "/usr/local/bin/sbcl"
          font-latex-fontify-script nil
          auto-save-default nil
          create-lockfiles nil
          delete-by-moving-to-trash t
          comment-column 64
          eshell-scroll-show-maximum-output nil

          ;; Change mode based on file extension
          auto-mode-alist (append '(("\\.cl$" . lisp-mode)
                                    ("\\.mod$" . go-mod-ts-mode))
                                  auto-mode-alist)
          ediff-split-window-function 'split-window-horizontally
          ediff-window-setup-function 'ediff-setup-windows-plain)

    ;; Reuse the inactive window instead of creating a new one.
    ;; Setting both split-height-threshold and split-width-threshold to nil seems to ensure this.
    ;; https://github.com/SophieBosio/.emacs.d
    (setq-default split-height-threshold nil
                  split-width-threshold  nil
                  fill-column            80) ;; Maximum line width

    ;; Rewrite the split-window-sensibly function
    ;; Reverse its preference and essentially prefer splitting side-by-side.
    ;; https://github.com/SophieBosio/.emacs.d
    (defun split-window-sensibly-prefer-horizontal (&optional window)
      "Based on `split-window-sensibly', but prefers to split WINDOW side-by-side."
      (let ((window (or window (selected-window))))
        (or (and (window-splittable-p window t)
                 ;; Split window horizontally
                 (with-selected-window window
                   (split-window-right)))
            (and (window-splittable-p window)
                 ;; Split window vertically
                 (with-selected-window window
                   (split-window-below)))
            (and
             ;; If WINDOW is the only usable window on its frame (it is
             ;; the only one or, not being the only one, all the other
             ;; ones are dedicated) and is not the minibuffer window, try
             ;; to split it horizontally disregarding the value of
             ;; `split-height-threshold'.
             (let ((frame (window-frame window)))
               (or
                (eq window (frame-root-window frame))
                (catch 'done
                  (walk-window-tree (lambda (w)
                                      (unless (or (eq w window)
                                                  (window-dedicated-p w))
                                        (throw 'done nil)))
                                    frame)
                  t)))
             (not (window-minibuffer-p window))
             (let ((split-width-threshold 0))
               (when (window-splittable-p window t)
                 (with-selected-window window
                   (split-window-right))))))))

    (defun split-window-really-sensibly (&optional window)
    (let ((window (or window (selected-window))))
      (if (> (window-total-width window) (* 2 (window-total-height window)))
          (with-selected-window window (split-window-sensibly-prefer-horizontal window))
        (with-selected-window window (split-window-sensibly window)))))
    (setq split-window-preferred-function 'split-window-really-sensibly)

    ;; Start with point at bottom when scratch buffers are created
    (add-hook 'emacs-startup-hook
          (lambda ()
            (when (get-buffer "*scratch*")
              (with-current-buffer "*scratch*"
                (goto-char (point-max))))))

    (setq-default indent-tabs-mode nil
                  global-tab-line-mode nil
                  tab-line-mode nil
                  tab-bar-mode nil
                  line-spacing 0.3
                  fill-column 80
                  sentence-end-double-space nil
                  whitespace-line-column 110
                  completion-at-point-functions (remove 'ispell-completion-at-point completion-at-point-functions)
                  org-catch-invisible-edits 'show
                  line-spacing (dot-env-get 'LINE_SPACING 1))

    ;; Do not show those confusing warnings when installing packages
    ;; https://protesilaos.com/codelog/2024-11-28-basic-emacs-configuration/
    (add-to-list 'display-buffer-alist
                 '("\\`\\*\\(Warnings\\|Compile-Log\\)\\*\\'"
                 (display-buffer-no-window)
                 (allow-no-window . t)))

    ;; Blink modeline instead of ring bell
    (setq ring-bell-function
          (lambda ()
            (let ((orig-fg (face-foreground 'mode-line)))
              (set-face-foreground 'mode-line "Magenta")
              (run-with-idle-timer 0.1 nil
                                   (lambda (fg) (set-face-foreground 'mode-line fg))
                                   orig-fg))))

    ;; Use forward slashes to distinguish duplicate files
    (require 'uniquify)
    (setq uniquify-buffer-name-style 'forward)

    ;; Maximize frames when created
    (add-to-list 'default-frame-alist '(fullscreen . maximized))

    (set-frame-font amo/font nil t)
    (set-face-attribute 'fixed-pitch nil :font amo/font)
    (set-face-attribute 'variable-pitch nil :font amo/font)
    (set-face-attribute 'default nil :height amo/base-font-size)

    (global-hl-line-mode t)
    (menu-bar-mode -1)

#+end_src

** Functions for custom bindings
#+begin_src emacs-lisp
  ;; Open file based from .emacs.d directory
  (defun amo/open-emacs-config-file (path)
    "Open a file within .emacs.d

  PATH is path to file inside .emacs.d"
    (interactive)
    (find-file (concat user-emacs-directory path)))

  ;; Exports README.org
  (defun amo/export-readme-file ()
    "Export README.org"
    (interactive)
    (let ((current-file-name (or (buffer-file-name) (buffer-name)))
          (is-file-p (buffer-file-name)))
      (find-file (concat user-emacs-directory "README.org"))
      (org-export-dispatch)
      (if is-file-p
          (find-file current-file-name)
        (switch-to-buffer current-file-name))))

  (defun amo/copy-rectangle-to-kill-ring (Begin End)
    "Copy region as column (rectangle region) to `kill-ring'
  (Because the native copy-rectangle-as-kill doesn't seem to work)
  See also: `kill-rectangle', `copy-to-register'.

  URL `http://xahlee.info/emacs/emacs/emacs_copy_rectangle_text_to_clipboard.html'
  Version: 2016-07-17"
    ;; extract-rectangle suggested by YoungFrog, 2012-07-25
    (interactive "r")
    (require 'rect)
    (kill-new (mapconcat #'identity (extract-rectangle Begin End) "\n")))

  ;; Un-highlight region after mark jump
  (defun amo/exchange-point-and-mark ()
    "Deactivates mark after exchanging point and mark"
    (interactive)
    (exchange-point-and-mark)
    (deactivate-mark))

  ;; Transpose chars and words backwards
  (defun amo/transpose-chars-backwards ()
    "Just like transpose-chars but goes the other way"
    (interactive)
    (transpose-chars -1))

  (defun amo/transpose-words-backwards ()
    "Just like transpose-words but goes the other way"
    (interactive)
    (transpose-words -1))

  ;; Move char after point to end of next word
  (defun amo/transpose-char-to-end-of-next-word ()
    "Move char at point to the end of the next word.
  Use case is to push closing parentheses out to contain subsequent characters
  when a function is typed and the closing parenthesis is automatically added.
  Skips over periods, quotes, and closing parentheses."
    (interactive)

    ;; helpers
    (defun amo/should-move-forward-one-char (c)
      (or
       (char-equal (following-char) ?\")
       (char-equal (following-char) ?\))))
    (defun amo/should-move-forward-one-word (c)
      (char-equal (following-char) ?.))
    (defun amo/should-move (c)
      (or
       (amo/should-move-forward-one-char c)
       (amo/should-move-forward-one-word c)))

    (forward-char)
    (let ((start (point))
          (end nil))
      (save-excursion
        (forward-word)
        (while (amo/should-move (following-char))
          (if (amo/should-move-forward-one-char (following-char))
              (forward-char)
            (forward-word)))
        (setq end (point)))
      (transpose-subr 'forward-char (- end start)))
    (backward-char))

  ;; Move lines up and down
  (defmacro amo/save-column (&rest body)
    `(let ((column (current-column)))
       (unwind-protect
           (progn ,@body)
         (move-to-column column))))
  (put 'save-column 'lisp-indent-function 0)

  (defun amo/move-line-up ()
    (interactive)
    (amo/save-column
     (transpose-lines 1)
     (forward-line -2)))

  (defun amo/move-line-down ()
    (interactive)
    (amo/save-column
     (forward-line 1)
     (transpose-lines 1)
     (forward-line -1)))

  ;; Toggle capitalization of character at point
  (defun amo/toggle-capitalization ()
    "Toggle the capitalization of the character at point."
    (interactive)
    (let ((char (char-after)))
      (cond ((eq char (upcase char)) (progn (amo/downcase-char 1) (forward-char)))
            ((eq char (downcase char)) (progn (upcase-char 1) (forward-char)))
            (t (message "No character at point.")))))

  (defun amo/downcase-char (arg)
    "Lowercasify ARG chars starting from point.  Point doesn't move."
    (interactive "p")
    (save-excursion
      (downcase-region (point) (progn (forward-char arg) (point)))))

  (defun amo/toggle-line-numbers ()
    "Toggle the display of line numbers"
    (interactive)
    (if (eq nil global-display-line-numbers-mode)
        (global-display-line-numbers-mode 1)
      (global-display-line-numbers-mode 0)))

  ;; Copy current file path & line number to kill ring
  (defun amo/copy-file-path-and-line-number (arg)
    "Copy current file path & line number to kill ring

  If called with the universal argument, provide full path.
  Otherwise use the projectile project root if present"
    (interactive "P")
    (let* ((file-path (or (buffer-file-name) (buffer-name)))
           (project-path
            (if (and (null arg) (projectile-project-root))
                (concat (file-name-nondirectory (directory-file-name (projectile-project-root)))
                        "/" (string-replace
                             (or (projectile-project-root) "") ""
                             file-path))
              file-path)))
      (kill-new (concat project-path
                        ":" (number-to-string (line-number-at-pos))))
      (message "Copied current path and line number to kill ring!")))

  ;; From https://github.com/larstvei/dot-emacs
  ;; just-one-space removes all whitespace around a point - giving it a negative argument it removes newlines as well.
  ;; We wrap a interactive function around it to be able to bind it to a key.
  ;; In Emacs 24.4 cycle-spacing was introduced, and it works like just-one-space,
  ;; but when run in succession it cycles between one, zero and the original number of spaces.
  (defun cycle-spacing-delete-newlines ()
    "Removes whitespace before and after the point."
    (interactive)
    (if (version< emacs-version "24.4")
        (just-one-space -1)
      (cycle-spacing -1)))

  (defun amo/repeat-current-window-right ()
    "Close other windows, duplicate current window to the right"
    (interactive)
    (delete-other-windows)
    (split-window-right)
    (other-window 1))

  (defun amo/repeat-current-window-below ()
    "Close other windows, duplicate current window below"
    (interactive)
    (delete-other-windows)
    (split-window-below)
    (other-window 1))

  ;; Reload config
  (defun amo/reload-config ()
    "Reload configuration"
    (interactive)
    (persp-state-save)
    (load-file (concat user-emacs-directory "init.el")))

  ;; Delete perspectives and customization file and restart emacs
  (defun amo/delete-emacs-perspective-file-restart ()
    "Delete saved desktop, then restart emacs"
    (interactive)
    (delete-file (concat user-emacs-directory "customize.el"))
    (persp-switch "main")
    (persp-kill-others)
    (restart-emacs))

  (defun amo/split-window (arg side)
    "Create a new window on SIDE.
    Universal argument will split based off root window.
    Double universal argument will split based off of window parent
    Also balance windows and move point to new window"
    (interactive "p")
    (split-window
     (cond
      ((= arg 4) (frame-root-window))
      ((= arg 16) (window-parent (selected-window)))
      ((= arg 1) (selected-window)))
     nil side nil)
    (balance-windows)
    (other-window 1))

  (defun amo/delete-window ()
    "Balance windows after deleting one"
    (interactive)
    (delete-window)
    (balance-windows))

  (defun amo/open-main-notes-file ()
    "Open main notes file"
    (interactive)
    (find-file (concat org-directory "/notes.org")))

  (defun amo/open-projects-file ()
    (interactive)
    (find-file (concat org-directory "/projects.org")))

  (defun amo/open-org-directory ()
    (interactive)
    (find-file org-directory))

  (defun amo/org-random-header-with-path ()
    "Pick a random header from the current Org buffer and display its full path."
    (interactive)
    (let ((header-paths '()))
      (org-map-entries
       (lambda ()
         (let ((full-path (amo/org-get-outline-path (org-get-heading t t t t))))
           (push full-path header-paths))) nil)
      (if header-paths
          (let* ((random-index (random (length header-paths)))
                 (random-header (nth random-index header-paths)))
            (message "Do This: %s" random-header))
        (message "No headers found."))))

  (defun amo/org-get-outline-path (header)
    "Get the full path of a HEADER in the Org buffer."
    (let ((path (list header)))
      (save-excursion
        (while (org-up-heading-safe)
          (push (org-get-heading t t t t) path)))
      (mapconcat 'identity path " -> ")))

#+end_src

** use-package Style Bindings
#+begin_src emacs-lisp
  :bind (("C-z r" . amo/reload-config)
         ("C-z C-z r" . restart-emacs)
         ("C-z C-z C-z r" . amo/delete-emacs-perspective-file-restart)
         ("C-z t" . consult-theme)
         ("C-z C-z t" . customize-themes)
         ("C-z e" . eshell)
         ("C-z n" . amo/open-main-notes-file)
         ("C-z i" . amo/open-projects-file)
         ("C-z C-z n" . amo/open-org-directory)
         ("C-z g" . amo/export-readme-file)
         ("C-z l" . amo/toggle-line-numbers)
         ("C-z p" . amo/copy-file-path-and-line-number)
         ("M-'" . end-of-visual-line)
         ("C-x C-x" . amo/exchange-point-and-mark)
         ("C-S-t" . amo/transpose-chars-backwards)
         ("M-T" . amo/transpose-words-backwards)
         ("C-S-f" . amo/transpose-char-to-end-of-next-word)
         ("C-:" . zap-up-to-char)
         ("M-@" . eval-expression)
         ("C-c t =" . text-scale-increase)
         ("C-c t -" . text-scale-decrease)
         ("C-z '" . indent-new-comment-line)
         ("C-x r M-w" . amo/copy-rectangle-to-kill-ring)
         ("M-/" . cycle-spacing-delete-newlines)
         ("C-c q" . remember)
         ("C-c Q" . remember-notes)
         ("C-c 1" . dictionary-lookup-definition)
         ("C-c 2" . amo/repeat-current-window-below)
         ("C-c 3" . amo/repeat-current-window-right)
         ("C-x 0" . amo/delete-window)
         ("M-<down>" . amo/move-line-down)
         ("M-<up>" . amo/move-line-up)
         ("C-'" . amo/toggle-capitalization)
         ("C-s-p" . scroll-down-line)
         ("C-s-n" . scroll-up-line)
         ("C-x K" . kill-buffer-and-window)
         ("M-SPC" . cycle-spacing)
         ("C-x t w" . tab-bar-mode)
         ))
#+end_src

** Traditional-Style Bindings
Necessary when the bound function requires arguments
#+begin_src emacs-lisp
  (global-set-key (kbd "C-x 2") (lambda (arg) (interactive "p") (amo/split-window arg 'below)))
  (global-set-key (kbd "C-x 3") (lambda (arg) (interactive "p") (amo/split-window arg 'right)))
  (global-set-key (kbd "C-z h") 'help-command)

  (keymap-global-set "C-z s e" (lambda () (interactive) (amo/open-emacs-config-file ".env")))
  (keymap-global-set "C-z s i" (lambda () (interactive) (amo/open-emacs-config-file "init.el")))
  (keymap-global-set "C-z s s" (lambda () (interactive) (amo/open-emacs-config-file "settings.org")))
  (keymap-global-set "C-z s t" (lambda () (interactive) (amo/open-emacs-config-file "terminal/settings.org")))
#+end_src
** Functions for hooks
#+begin_src emacs-lisp
  ;; Don't auto-complete when we are debugging
  (defun amo/comint-mode-actions ()
    (setq-local corfu-auto nil) [])

  ;; Cleanup whitespace
  (defun amo/whitespace-cleanup ()
    (whitespace-cleanup)
    (delete-trailing-whitespace))
  #+end_src

** Hooks
#+begin_src emacs-lisp
  (add-hook 'before-save-hook 'amo/whitespace-cleanup)
  (add-hook 'text-mode-hook 'flyspell-mode)
  (add-hook 'prog-mode-hook (lambda () (flyspell-mode -1)))
  (add-hook 'prog-mode-hook 'subword-mode)
  (add-hook 'calc-mode-hook (lambda () (setq line-spacing nil)))
  (add-hook 'calc-trail-mode-hook (lambda () (setq line-spacing nil)))
  (add-hook 'comint-mode-hook 'amo/comint-mode-actions)
  (add-hook 'focus-out-hook 'garbage-collect)

  ;; Save theme that consult-theme may have selected but not saved
  (require 'cus-theme)
  (add-hook 'kill-emacs-hook 'custom-theme-save)
  (add-hook 'restart-emacs-hook 'custom-theme-save)
#+end_src
** dired
#+begin_src emacs-lisp
  (use-package dired :ensure nil
    :config (setq dired-kill-when-opening-new-dired-buffer t
                  dired-listing-switches "-ahl --group-directories-first")
    :hook ((dired-mode . dired-hide-details-mode))
    :bind (:map dired-mode-map
                ("C-o" . nil)))
#+end_src

** ibuffer
#+begin_src emacs-lisp
  (use-package ibuffer
    ;; So it doesn't clobber ace-window's binding
    :bind (:map ibuffer-mode-map
                ("C-o" . nil)))
#+end_src

** Advice
#+begin_src emacs-lisp
  ;; Copy whole line to kill ring when no active region
  (defadvice kill-ring-save (before slick-copy activate compile)
    "When called interactively with no active region, copy a single line instead."
    (interactive
     (if mark-active
         (list (region-beginning) (region-end))
       (message "Copied line")
       (list (line-beginning-position) (line-beginning-position 2)))))
#+end_src

* Packages
** Minibuffer & Completion
*** Vertico
- Repo :: https://github.com/minad/vertico
- Description :: Performant and minimalistic vertical completion UI based on the default completion system.
#+begin_src emacs-lisp
  (use-package vertico
    :ensure (:files (:defaults "extensions/*"))
    :bind (:map vertico-map
                ("C-j" . vertico-directory-enter)
                ("DEL" . vertico-directory-delete-char)
                ("M-DEL" . vertico-directory-delete-word))
    :hook (rfn-eshadow-update-overlay . vertico-directory-tidy)
    :init
    (vertico-mode)
    :config
    (setq vertico-resize t)
    (advice-add 'vertico--setup :before (lambda () (setq vertico-count 10)))
    (define-key minibuffer-local-map (kbd "s-'") (lambda ()
                                                   (interactive)
                                                   (setq vertico-count (if (= vertico-count 10)
                                                                           (- (floor  (* (frame-height) 0.75)) 10)
                                                                         10))
                                                   (vertico--exhibit))))
#+end_src
*** Embark
- Repo :: https://github.com/oantolin/embark
- Description :: Emacs Mini-Buffer Actions Rooted in Keymaps
#+begin_src emacs-lisp
(use-package embark
  :ensure t
  :bind (("C-." . embark-act)
         ("C-S-h B" . embark-bindings)
         :map flyspell-mode-map
         ("C-." . nil)
         :map occur-mode-map
         ("C-c C-p" . amo/convert-occur-to-wgrep))

  :init
  (setq prefix-help-command #'embark-prefix-help-command)

  :config

terminal/settings.el
;; A minimal terminal emacs config
;; Add `alias emacs="emacs -nw -q --load '~/.emacs.d/terminal/init.el'"`
;; to your shell configuration

(straight-use-package
 '(nerd-icons :type git :host github :repo "rainstormstudio/nerd-icons.el"))

(setq inhibit-startup-screen t
      save-interprogram-paste-before-kill t
      x-select-enable-clipboard t
      auto-save-default nil
      make-backup-files nil
      create-lockfiles nil)

(setq-default indent-tabs-mode nil
              global-tab-line-mode nil
              tab-line-mode nil
              tab-bar-mode nil
              line-spacing 0.3
              fill-column 100
              sentence-end-double-space nil
              visual-line-mode t
              whitespace-line-column 110)

(add-hook 'before-save-hook 'whitespace-cleanup)

(global-display-line-numbers-mode 0)
(fset 'yes-or-no-p 'y-or-n-p)
(tool-bar-mode -1)
(menu-bar-mode -1)
(blink-cursor-mode 0)
(electric-pair-mode 1)
(delete-selection-mode 1)
(recentf-mode 1)
(savehist-mode 1)
(whitespace-mode -1)
(define-prefix-command 'z-map)

(defadvice kill-ring-save (before slick-copy activate compile)
  "When called interactively with no active region, copy a single line instead."
  (interactive
   (if mark-active
       (list (region-beginning) (region-end))
     (message "Copied line")
     (list (line-beginning-position) (line-beginning-position 2)))))

;; Use C-h for delete-backward-char
(global-set-key (kbd "C-h") 'delete-backward-char)

(global-set-key (kbd "C-z") 'z-map)

(defun amo/modus-themes-toggle (&optional ignore-if-set)
  "Toggle modus theme, sets to operandi if unset"
  (interactive)
  (if (string-match "modus" (symbol-name (car custom-enabled-themes)))
      (when (not ignore-if-set)
        (modus-themes-toggle))
    (load-theme 'modus-operandi)))
(amo/modus-themes-toggle t)
(define-key z-map (kbd "t") 'amo/modus-themes-toggle)

(defun amo/open-settings-file ()
  "Open settings.el"
  (interactive)
  (find-file "~/.emacs.d/terminal/settings.el"))
(define-key z-map (kbd "s") 'amo/open-settings-file)

(setq ring-bell-function
      (lambda ()
        (let ((orig-fg (face-foreground 'mode-line)))
          (set-face-foreground 'mode-line "Magenta")
          (run-with-idle-timer 0.1 nil
                               (lambda (fg) (set-face-foreground 'mode-line fg))
                               orig-fg))))

;; Reload config
(defun amo/reload-config ()
  "Reload configuration"
  (interactive)
  ;; (dot-env-load)
  (load-file "~/.emacs.d/terminal/init.el"))
(define-key z-map (kbd "r") #'amo/reload-config)

;; Toggle line numbers
(defun amo/toggle-line-numbers ()
  "Toggle the display of line numbers"
  (interactive)
  (if (eq nil global-display-line-numbers-mode)
      (global-display-line-numbers-mode 1)
    (global-display-line-numbers-mode 0)))
(define-key z-map (kbd "l") 'amo/toggle-line-numbers)

;; Toggle capitalization of character at point
(defun amo/toggle-capitalization ()
  "Toggle the capitalization of the character at point."
  (interactive)
  (let ((char (char-after)))
    (cond ((eq char (upcase char)) (progn (amo/downcase-char 1) (forward-char)))
          ((eq char (downcase char)) (progn (upcase-char 1) (forward-char)))
          (t (message "No character at point.")))))

(defun amo/downcase-char (arg)
  "Lowercasify ARG chars starting from point.  Point doesn't move."
  (interactive "p")
  (save-excursion
    (downcase-region (point) (progn (forward-char arg) (point)))))

;; Move lines up and down
(defmacro amo/save-column (&rest body)
  `(let ((column (current-column)))
     (unwind-protect
         (progn ,@body)
       (move-to-column column))))
(put 'save-column 'lisp-indent-function 0)

(defun amo/move-line-up ()
  (interactive)
  (amo/save-column
   (transpose-lines 1)
   (forward-line -2)))

(defun amo/move-line-down ()
  (interactive)
  (amo/save-column
   (forward-line 1)
   (transpose-lines 1)
   (forward-line -1)))

(put 'narrow-to-region 'disabled nil)

(use-package ace-window
  :straight t
  :bind
  (("C-o" . ace-window))
  :config
  (defvar aw-dispatch-always nil)
  (setq aw-scope 'frame)
  (custom-set-faces
   '(aw-leading-char-face
     ((t (:inherit ace-jump-face-foreground :height 3.0))))))

(use-package aggressive-indent
    :straight t
    :config
    (global-aggressive-indent-mode 1)
    (add-to-list 'aggressive-indent-excluded-modes 'html-mode)
    (add-to-list 'aggressive-indent-excluded-modes 'python-ts-mode))

(use-package avy
  :straight t)

(use-package company
  :straight t
  :config
  (global-company-mode))

(use-package crux
  :straight t
  :bind
  (("M-o" . crux-smart-open-line)
   ("C-c D" . crux-delete-file-and-buffer)
   ("C-c d" . crux-duplicate-current-line-or-region)
   ("C-c M-d" . crux-duplicate-and-comment-current-line-or-region)
   ("C-c r" . crux-rename-file-and-buffer)
   ("<C-backspace>" . crux-kill-line-backwards)))

(use-package expand-region
  :straight t)

(use-package vertico
  :straight (:files (:defaults "extensions/*"))
  :bind (:map vertico-map
              ("C-j" . vertico-directory-enter)
              ("DEL" . vertico-directory-delete-char)
              ("M-DEL" . vertico-directory-delete-word))
  :hook (rfn-eshadow-update-overlay . vertico-directory-tidy)
  :init
  (vertico-mode))

(use-package marginalia
  :straight t
  :init
  (marginalia-mode))

(use-package consult
  :straight t

  ;; Replace bindings. Lazily loaded due by `use-package'.
  :bind (("C-s" . consult-line)
         ("C-c m" . consult-man)
         ("C-c i" . consult-info)
         ([remap Info-search] . consult-info)
         ("C-x b" . consult-buffer)
         ("C-x 4 b" . consult-buffer-other-window)
         ("C-x r b" . consult-bookmark)
         ("M-y" . consult-yank-pop)
         ("M-g g" . consult-goto-line)
         ("M-g o" . consult-outline)
         ("M-g m" . consult-mark)
         ("M-g k" . consult-global-mark)
         ("M-g i" . consult-imenu)
         ("M-g I" . consult-imenu-multi)
         ("M-s d" . consult-find)
         ("M-s D" . consult-locate)
         ("M-s g" . consult-grep)
         ("M-s G" . consult-git-grep)
         ("M-s r" . consult-ripgrep)
         ("M-s e" . consult-isearch-history))

  ;; Enable automatic preview at point in the *Completions* buffer. This is
  ;; relevant when you use the default completion UI.
  :hook (completion-list-mode . consult-preview-at-point-mode)

  ;; The :init configuration is always executed (Not lazy)
  :init

  ;; Optionally configure the register formatting. This improves the register
  ;; preview for `consult-register', `consult-register-load',
  ;; `consult-register-store' and the Emacs built-ins.
  (setq register-preview-delay 0.5
        register-preview-function #'consult-register-format)

  ;; Optionally tweak the register preview window.
  ;; This adds thin lines, sorting and hides the mode line of the window.
  (advice-add #'register-preview :override #'consult-register-window)

  ;; Use Consult to select xref locations with preview
  (setq xref-show-xrefs-function #'consult-xref
        xref-show-definitions-function #'consult-xref)

  ;; Configure other variables and modes in the :config section,
  ;; after lazily loading the package.
  :config
  (consult-customize
   consult-theme :preview-key '(:debounce 0.2 any)
   consult-ripgrep consult-git-grep consult-grep
   consult-bookmark consult-recent-file consult-xref
   consult--source-bookmark consult--source-file-register
   consult--source-recent-file consult--source-project-recent-file
   ;; :preview-key "M-."
   :preview-key '(:debounce 0.4 any)))


(use-package orderless
  :straight t
  :custom
  (completion-styles '(orderless basic))
  (completion-category-overrides '((file (styles basic partial-completion)))))

(use-package nerd-icons-completion
  :straight t
  :after marginalia
  :hook ((marginalia-mode . nerd-icons-completion-marginalia-setup))
  :init
  (nerd-icons-completion-mode 1))

(use-package which-key
  :straight t
  :config
  (which-key-mode))

(use-package magit
  :straight t
  :bind
  (("C-x g" . magit)))

(use-package jump-char
  :straight (:host github :repo "lewang/jump-char"
                   :branch "master")
  :bind (("M-n" . jump-char-forward)
         ("M-N" . jump-char-backward)))

(use-package idle-highlight-mode
  :straight t
  :config
  (setq idle-highlight-idle-time 0.2
        idle-highlight-exclude-point t)
  :hook
  ((prog-mode text-mode) . idle-highlight-mode))

(use-package minions
  :straight t
  :config
  (minions-mode 1))

(use-package paredit
  :straight t
  :hook
  ((lisp-mode . paredit-mode)
   (emacs-lisp-mode . paredit-mode)
   (clojure-mode . paredit-mode)
   (clojurescript-mode . paredit-mode)
   (clojurec-mode . paredit-mode)
   (cider-repl-mode . paredit-mode)))

(use-package undo-tree
  :straight t
  :config
  (global-undo-tree-mode)
  (setq undo-tree-history-directory-alist `((".*" . ,temporary-file-directory))
        undo-tree-enable-undo-in-region t
        undo-tree-auto-save-history t)
  :diminish
  (undo-tree-mode))

;; Terminal mappings to support iterm2 for Mac
;; https://www.emacswiki.org/emacs/iTerm2#h5o-11
(defun amo/define-escape-sequences-and-bindings ()
  (interactive)
  (progn
    (let ((map (if (boundp 'input-decode-map)
                   input-decode-map
                 function-key-map)))

      ;; C-=
      (define-key map "\e[1;P9"  (kbd "H-a"))
      (global-set-key (kbd "H-a") 'er/expand-region)

      ;; C--
      (define-key map "\e[1;P10" (kbd "H-b"))
      (global-set-key (kbd "H-b") 'er/contract-region) ;; FIXME - maps to undo for some reason

      ;; C-;
      (define-key map "\e[1;P11" (kbd "H-c"))
      (global-set-key (kbd "H-c") 'avy-goto-char-2)
      (global-set-key (kbd "C-x H-c") 'comment-line)

      ;; s-o
      (define-key map "\e[1;P12" (kbd "H-d"))
      (global-set-key (kbd "H-d") 'crux-smart-open-line-above)

      ;; s-j
      (define-key map "\e[1;P13" (kbd "H-e"))
      (global-set-key (kbd "H-e") 'crux-top-join-line)

      ;; s-k
      (define-key map "\e[1;P14" (kbd "H-f"))
      (global-set-key (kbd "H-f") 'crux-kill-whole-line)

      ;; C-s-p
      (define-key map "\e[1;P15" (kbd "H-g"))
      (global-set-key (kbd "H-g") 'scroll-down-line)

      ;; C-s-n
      (define-key map "\e[1;P16" (kbd "H-h"))
      (global-set-key (kbd "H-h") 'scroll-up-line)

      ;; C-S-h
      (define-key map "\e[1;P17" (kbd "H-i"))
      (global-set-key (kbd "H-i") 'help-command)

      ;; M-S<down>
      (define-key map "\e[1;P18" (kbd "H-j"))
      (global-set-key (kbd "H-j") 'amo/move-line-down)

      ;; M-S-<up>
      (define-key map "\e[1;P19" (kbd "H-k"))
      (global-set-key (kbd "H-k") 'amo/move-line-up)

      ;; C-'
      (define-key map "\e[1;P20" (kbd "H-l"))
      (global-set-key (kbd "H-l") 'amo/toggle-capitalization)

      ;; Additional available keybindings
      (define-key map "\e[1;P21" (kbd "H-m"))
      (define-key map "\e[1;P22" (kbd "H-n"))
      (define-key map "\e[1;P23" (kbd "H-o"))
      (define-key map "\e[1;P24" (kbd "H-p"))
      (define-key map "\e[1;P25" (kbd "H-q"))
      (define-key map "\e[1;P26" (kbd "H-r"))
      (define-key map "\e[1;P27" (kbd "H-s"))
      (define-key map "\e[1;P28" (kbd "H-t"))
      (define-key map "\e[1;P29" (kbd "H-u"))
      (define-key map "\e[1;P30" (kbd "H-v"))
      (define-key map "\e[1;P31" (kbd "H-w"))
      (define-key map "\e[1;P32" (kbd "H-x"))
      (define-key map "\e[1;P33" (kbd "H-y"))
      (define-key map "\e[1;P34" (kbd "H-z"))
      (define-key map "\e[1;P35" (kbd "H-0"))
      (define-key map "\e[1;P36" (kbd "H-1"))
      (define-key map "\e[1;P37" (kbd "H-2"))
      (define-key map "\e[1;P38" (kbd "H-3"))
      (define-key map "\e[1;P39" (kbd "H-4"))
      (define-key map "\e[1;P40" (kbd "H-5"))
      (define-key map "\e[1;P41" (kbd "H-6"))
      (define-key map "\e[1;P42" (kbd "H-7"))
      (define-key map "\e[1;P43" (kbd "H-8"))
      (define-key map "\e[1;P44" (kbd "H-9"))
      (define-key map "\e[1;P45" (kbd "H-<f1>"))
      (define-key map "\e[1;P46" (kbd "H-<f2>"))
      (define-key map "\e[1;P47" (kbd "H-<f3>"))
      (define-key map "\e[1;P48" (kbd "H-<f4>"))
      (define-key map "\e[1;P49" (kbd "H-<f5>"))
      (define-key map "\e[1;P50" (kbd "H-<f6>"))
      (define-key map "\e[1;P51" (kbd "H-<f7>"))
      (define-key map "\e[1;P52" (kbd "H-<f8>"))
      (define-key map "\e[1;P53" (kbd "H-<f9>"))
      (define-key map "\e[1;P54" (kbd "H-<f10>"))
      (define-key map "\e[1;P55" (kbd "H-<f11>"))
      (define-key map "\e[1;P56" (kbd "H-<f12>")))))

(add-hook 'server-after-make-frame-hook 'amo/define-escape-sequences-and-bindings)
